onnx/shape_inference/implementation.h

```diff
@@ -0,0 +1,135 @@
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+void InferShapes(ModelProto& m) {
+  auto* g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor *> valueTypesByName;
+  for (auto& vi : *g->mutable_value_info()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_input()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_output()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+
+  for (const auto& n : g->node()) {
+    auto schema = OpSchemaRegistry::Schema(n.op_type());
+    if (schema) {
+      InferenceContextImpl ctx;
+
+      {
+        std::unordered_map<std::string, const AttributeProto *> attributesByName;
+        for (const auto& attr : n.attribute()) {
+          attributesByName[attr.name()] = &attr;
+        }
+        ctx.attributesByName_ = std::move(attributesByName);
+      }
+
+      {
+        std::vector<const TypeProto_Tensor *> allInputTypes;
+
+        for (const auto& input : n.input()) {
+          auto iter = valueTypesByName.find(input);
+          if (iter != valueTypesByName.end()) {
+            allInputTypes.push_back(iter->second);
+          } else {
+            allInputTypes.push_back(nullptr);
+          }
+        }
+
+        ctx.allInputTypes_ = std::move(allInputTypes);
+      }
+
+      ctx.allOutputTypes_.resize(n.output_size());
+
+      schema->GetShapeInferenceFunction()(ctx);
+
+      for (int i = 0; i < n.output_size(); i++) {
+        const auto& output = n.output(i);
+        const auto& inferredType = *ctx.getOutputType(i);
+
+        // In this case, we have no new information, so don't bother
+        // to add a contentless ValueInfo.
+        if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+            !inferredType.has_shape()) {
+          continue;
+        }
+
+        // If there is already a ValueInfo associated with this
+        // output, reuse it. Otherwise add a new one.
+        auto iter = valueTypesByName.find(output);
+        TypeProto_Tensor* existingType = nullptr;
+        if (iter != valueTypesByName.end()) {
+          existingType = iter->second;
+        } else {
+          auto vi = g->add_value_info();
+          vi->set_name(output);
+          existingType = vi->mutable_type()->mutable_tensor_type();
+        }
+
+        // Incorporate the inferred information.
+        if (inferredType.elem_type() != TensorProto::UNDEFINED) {
+          if (existingType->elem_type() != TensorProto::UNDEFINED &&
+              existingType->elem_type() != inferredType.elem_type()) {
+            throw std::runtime_error("inferred type differs from existing type");
+          } else {
+            existingType->set_elem_type(inferredType.elem_type());
+          }
+        }
+        for (int j = 0; j < inferredType.shape().dim_size(); j++) {
```

Comment:
hm, I think you actually need to error out if the sizes are different. I.e. if ranks conflict

Reply:
yeah, i guess I am working around some bugs in make_tensor_value_info() on the python side - it still uses empty dimensions to represent a missing shape. I'll fix that code and then I can just use has_shape() here